api/http.go

```diff
@@ -1,69 +1,58 @@
 package api
 
 import (
-	"fmt"
+	"context"
 	"net/http"
+	"time"
 
 	"github.com/julienschmidt/httprouter"
+	"github.com/livepeer/catalyst-api/balancer"
+	"github.com/livepeer/catalyst-api/cluster"
 	"github.com/livepeer/catalyst-api/config"
-	"github.com/livepeer/catalyst-api/handlers"
-	"github.com/livepeer/catalyst-api/handlers/misttriggers"
+	"github.com/livepeer/catalyst-api/handlers/geolocation"
 	"github.com/livepeer/catalyst-api/log"
 	"github.com/livepeer/catalyst-api/middleware"
-	"github.com/livepeer/catalyst-api/pipeline"
 )
 
-func ListenAndServe(apiPort int, apiToken string, vodEngine *pipeline.Coordinator) error {
-	listen := fmt.Sprintf("0.0.0.0:%d", apiPort)
-	router := NewCatalystAPIRouter(vodEngine, apiToken)
+func ListenAndServe(ctx context.Context, cli config.Cli, bal balancer.Balancer, c cluster.Cluster) error {
+	router := NewCatalystAPIRouter(cli, bal, c)
+	server := http.Server{Addr: cli.HTTPAddress, Handler: router}
+	ctx, cancel := context.WithCancel(ctx)
 
 	log.LogNoRequestID(
 		"Starting Catalyst API!",
 		"version", config.Version,
-		"host", listen,
+		"host", cli.HTTPAddress,
 	)
-	return http.ListenAndServe(listen, router)
-}
 
-func NewCatalystAPIRouter(vodEngine *pipeline.Coordinator, apiToken string) *httprouter.Router {
-	router := httprouter.New()
-	withLogging := middleware.LogRequest()
-	withAuth := middleware.IsAuthorized
-	withCapacityChecking := middleware.HasCapacity
+	var err error
+	go func() {
+		err = server.ListenAndServe()
+		cancel()
+	}()
 
-	catalystApiHandlers := &handlers.CatalystAPIHandlersCollection{VODEngine: vodEngine}
-	mistCallbackHandlers := &misttriggers.MistCallbackHandlersCollection{VODEngine: vodEngine}
+	<-ctx.Done()
```

Comment:
Ah yeah — we still need something monitoring the context to shut down the server if necessary, but it's probably cleaner to put that part in the goroutine.

Reply:
Actually thinking about it more, I think I was concerned about the case where `Shutdown` doesn't stop the server in time; maybe it's stuck for some reason. In that case we do want to return early so that the application can continue shutting down, rather than blocking the shutdown process forever.